geowebcache/core/src/main/java/org/geowebcache/layer/wms/WMSHttpHelper.java

```diff
@@ -173,7 +173,17 @@ private void connectAndCheckHeaders(
         try {
             method = executeRequest(wmsBackendUrl, wmsParams, backendTimeout, httpRequestMode);
             responseCode = method.getStatusLine().getStatusCode();
-            responseLength = Integer.parseInt(method.getFirstHeader("length").getValue());
+            if (method.getFirstHeader("length") != null) {
+                responseLength = Integer.parseInt(method.getFirstHeader("length").getValue());
+            } else if (method.getFirstHeader("Content-Length") != null) {
+                responseLength =
+                        Integer.parseInt(method.getFirstHeader("Content-Length").getValue());
+            } else if (method.getEntity() != null) {
+                responseLength = Math.toIntExact(method.getEntity().getContentLength());
+            } else {
+                throw new ServiceException(
```

Comment:
Never mind, method.getEntity() should cover that case, when we get in that branch the content lenght can still be 0 if not known, or even negative. So the last branch throwing exception should be reached only if the server did not provide a body altogether.

Which brings me to another consideration: the status code should be checked before we complain about the content lenght, otherwise we complain that we cannot find out the content length, while the status was telling using something like "forbidden" or "internal server error"... the exception would hide the actual reason of the lack of a body.

Reply:
Fixed